qiskit/pulse/instructions/instruction.py

```diff
@@ -219,6 +219,37 @@ def append(self, schedule: ScheduleComponent,
         time = self.ch_stop_time(*common_channels)
         return self.insert(time, schedule, name=name)
 
+    def assign_parameters(self,
+                          value_dict: Dict[ParameterExpression,
+                                           Union[ParameterExpression, int, float, complex]],
+                          inplace: bool = True) -> 'Instruction':
+        """Assign the parameters in this instruction according to the input.
+
+        Args:
+            value_dict: A mapping from Parameters to either numeric values or another
+                Parameter expression.
+            inplace: Modify this ``Instruction`` iff True, else return a new ``Instruction``.
+
+        Returns:
+            Instruction with updated parameters (a new one if not inplace, otherwise self).
+        """
+        new_operands = list(self.operands)
+        for idx, op in enumerate(self.operands):
+            for parameter, value in value_dict.items():
+                if isinstance(op, ParameterExpression) and parameter in op.parameters:
+                    new_operands[idx] = new_operands[idx].assign(parameter, value)
+                elif (isinstance(op, Channel)
+                      and isinstance(op.index, ParameterExpression)
+                      and parameter in op.index.parameters):
+                    # TODO: Check int type?
+                    new_index = new_operands[idx].index.assign(parameter, value)
+                    new_operands[idx] = type(op)(new_index)
+
+        if inplace:
+            self._operands = tuple(new_operands)
```

Comment:
I'm starting to wonder if we should just switch everything to mutable and request that if they want to do an immutable operation just request they copy and then mutate, eg., `new_inst = copy.copy(inst).assign_parameter(...)`. It might be very cumbersome to always maintain `inplace` going forward. What do you think?

Reply:
I definitely agree. I can take that approach here (although, ironically and confusingly, this is one of the only immutable operation in circuits).
Generally, it was very cumbersome to change the default behavior for all operations and the performance was an issue

Reply:
@taalexander Do you mean this for schedule and pulse, too?